### PR TITLE
Add match-field-type property in releated

### PR DESCRIPTION
**Pull request by minhuyen**
I added match-field-type property support 2 options: string vs ObjectId in related config. The default value is ObjectId.
It supports convert src-field to string if the match-field-type value is a string. 
The pull request support for #464 issue

### CODE REVIEW

**Comment by rwynn** [comment]
Hi @minhuyen thanks for your contribution to monstache.  Can we work on this PR together a little before merging the feature?  

My suggestions would be:

1. Assume no default type for the `srcData` and thus if the user does not specify `match-field-type` then keep it as empty string instead of defaulting to `ObjectID`
2. Extract the actual data conversion into a separate function.  In this function support converting srcData from `string` to `objectId` (via primitive.ObjectIDFromHex) or `objectId` to `string` like you already added depending on the value of `match-field-type`.  Use the exact `alias` values from https://docs.mongodb.com/manual/reference/bson-types/ as the accepted inputs for `match-field-type`.  Have the function return an error in the case where the conversion fails.  
3. Add a few tests to monstache_test.go to cover conversion from `string` to `objectId` or vice versa and also some error cases. 

Thanks,
Ryan

**Comment by minhuyen** [comment]
Hi @rwynn 
I have updated the code according to your suggestion.
Please check! Thanks

**Comment by rwynn** [comment]
Hi @minhuyen thanks, I like the changes you have made so far.  

Can you please update your branch so that the changes listed on this PR (see Files changed tab) are only those related to this feature?  I think you may need to merge your branch with the latest from the `rel6` branch.  

That will make it easier to discuss.  Thanks!

**Comment by minhuyen** [comment]
Hi @rwynn 
I updated my branch. Can you check again?
Thanks.

File: monstache.go [hunk 1]
```diff
@@ -198,6 +198,7 @@ type relation struct {
 	DotNotation   bool   `toml:"dot-notation"`
 	KeepSrc       bool   `toml:"keep-src"`
 	MaxDepth      int    `toml:"max-depth"`
+	MatchFieldType string `toml:"match-field-type"`
```
Comment (rwynn) [inline monstache.go:201]: Can you run `gofmt` on this file?

File: monstache.go [hunk 2]
```diff
@@ -1070,6 +1071,27 @@ func buildSelector(matchField string, data interface{}) bson.M {
...
+	case primitive.ObjectID:
+		value = srcData.(primitive.ObjectID).Hex()
+	default:
+		value = fmt.Sprintf("%v", v)
+	}
+	return
+}
+
+func convertSrcDataToObjectId(srcData interface{}) (objectId primitive.ObjectID, err error) {
+	defer func() {
```
Comment (rwynn) [inline monstache.go]: I don't think the defer and recover is necessary here.

File: monstache.go [hunk 3]
```diff
@@ -1107,6 +1129,15 @@ func (ic *indexClient) processRelated(root *gtm.Op) (err error) {
 					continue
 				}
 
+				if r.MatchFieldType == "objectId" {
```
Comment (rwynn) [inline monstache.go]: Can you extract this part into one more function, e.g. `convertSrcDataToMatchFieldType` like

```golang
if r.MatchFieldType != "" {
if srcData, err = convertSrcDataToMatchFieldType(srcData, r.MatchFieldtype); err != nil {
						ic.processErr(err)
						continue
					}
}
```
This new function `convertSrcDataToMatchFieldType` can then check which of `convertSrcDataToObjectId` or `convertSrcDataToString` to invoke.

**Review by rwynn — CHANGES_REQUESTED** [review]

**Comment by minhuyen** [comment]
Hi, @rwynn thanks for your feedback!

I updated the code. Can you review it?

Thanks

File: monstache.go [hunk 4]
```diff
@@ -1070,6 +1071,31 @@ func buildSelector(matchField string, data interface{}) bson.M {
 	return sel
 }
 
+func convertSrcDataToString(srcData interface{}) (value string) {
+	switch v := srcData.(type) {
+	case primitive.ObjectID:
+		value = srcData.(primitive.ObjectID).Hex()
```
Comment (rwynn) [inline monstache.go]: Can you change this line to `v.Hex()` since v will already be ObjectID here?

**Review by rwynn — CHANGES_REQUESTED** [review]
This is looking really good.  Thanks for your patience and working with me on it.

Just had one last request that I forgot to add to the last review.  Then I think I'm good to merge it.

**Comment by minhuyen** [comment]
I have updated it.
Thanks for your support

**Review by rwynn — APPROVED** [review]
Thanks for contributing this feature!

**Pull request merged**